### DISCUSSION TITLE
How is printf auto-static yoinking of symbols supposed to interact with things like function pointers ?

### DISCUSSION THREAD

**Discussion by GabrielRavier** [jart/cosmopolitan#797] [General]
Having spent a little while studying how the auto-linking in printf works, I'd like to ask if the fact that it simply doesn't work with function pointers is deliberate behavior or effectively a bug.

In particular, a program such as this:

```cpp
#include <stdio.h>

void use_printf_through_function_pointer (int (*printf_ptr)(const char *, ...))
{
    printf_ptr("%f\n", 12.75);
}

int
main (void)
{
    use_printf_through_function_pointer(printf);
    return 0;
}
```

simply doesn't work under cosmopolitan because it never links in `__fmt_dtoa`, as the invocation of `printf_ptr` cannot be scanned for that purpose. That appears to be deliberate (at least, blog posts such as [this one](https://justine.lol/sizetricks/) directly indicate that non-direct invocations of printf disable the auto-`STATIC_YOINK` mechanism), but I'm wondering if there's any way of disabling that behavior, through a command-line switch to gcc when compiling or something like that.

In particular, I'm asking this because I'm currently developing a libc testsuite [over here](https://github.com/GabrielRavier/yalibct/) (that's why I've filed 17 bug reports in just a few days), and tests such as [this](https://github.com/GabrielRavier/yalibct/blob/master/tests/printf/gnulib-test-posix.c) [one](https://github.com/GabrielRavier/yalibct/blob/master/tests/printf/gnulib-test-posix.h) explicitly test that printf works properly when called through a function pointer.

Right now it's not too much hassle to add in a

```cpp
#ifdef __COSMOPOLITAN__
STATIC_YOINK("__fmt_dtoa");
// etc.
#endif
```

but it feels a bit weird to do so, especially as this still means perfectly standards-conforming programs still fail to run properly unless the manual `STATIC_YOINK` is present.

I suppose that could just be added to the list of ways in which cosmopolitan deliberately deviates from the standard, but for this in particular it seems odd that there is no way of disabling this behavior.

**Comment by G4Vi** [comment]
I haven't tried this, you may be able to force `gcc` to link the necessary symbols with the link option `-u symbol`. For example, `-u __fmt_dtoa`.

The only official cosmopolitan method I know of is building with `MODE=ansi`: https://github.com/jart/cosmopolitan/blob/f7ff77d865ca9f55d02abfea71573baedd1f188d/build/config.mk#L365-L397 When building with `__STRICT_ANSI__`, `_weaken` is set as follows: `#define _weaken(symbol)      symbol` https://github.com/jart/cosmopolitan/blob/6f7d0cb1c3962f7cb474b04de75df7cd82a3e5d3/libc/intrin/weaken.h

**Reply by GabrielRavier** [reply]
Ah, nice, I guess that's probably a much better configuration to test conformance on than `MODE=dbg` (although that means it doesn't have all the debugging stuff, but adding `ansidbg` should be trivial)

**Reply by GabrielRavier** [reply]
...The part where `MODE=ansi` just doesn't compile is pretty unhelpful, though.

**Reply by GabrielRavier** [reply]
Is `MODE=ansi` something that's supposed to be capable of working with cosmopolitan as a whole ? The amount of build errors is humongous and it seems like something that'd take hours, if not days, to fix up...

**Reply by G4Vi** [reply]
I'd assume it's supposed to be work, but I don't think there's been a persistent effort to maintain it. I haven't tried a `MODE=ansi` build until today and the CI doesn't build it.

It may be helpful to discuss your libc test suite and cosmopolitan in our discord server, link if you haven't joined: https://discord.gg/7FZGecjP Other collaborators or discord members may have a better idea of what works or doesn't work with `MODE=ansi` or how we may be able to accommodate your efforts. I greatly appreciate all the discoveries and fixes you've made on our format string functionality.

**Reply by jart** [reply]
Yeah I haven't used `MODE=ansi` in the longest time. IIRC it was only ever good for making the headers? Don't remember.
